server/src/hicstat.ts

```diff
@@ -112,8 +112,127 @@ export async function do_hicstat(file: string, isurl: boolean): Promise<HicstatR
 		})
 	//console.log('Reading matrix ...')
 	out_data.normalization = normalization
+
+	// Fetch observed and expected values for all resolutions
+	const resolutions = out_data['Base pair-delimited resolutions'].concat(out_data['Fragment-delimited resolutions'])
+
+	for (const resolution of resolutions) {
+		const observedExpectedData: Record<string, ResolutionData> = {}
+
+		for (const chr1 of out_data.chrorder) {
+			for (const chr2 of out_data.chrorder) {
+				const observed = await getObservedValues(file, resolution, chr1, chr2)
+				const expected = await getExpectedValues(file, resolution, chr1, chr2)
+				const observedOverExpected = calculateObservedOverExpected(observed, expected)
+
+				// Store the data in resolutionsData
+				observedExpectedData[`${chr1}-${chr2}`] = {
+					observed,
+					expected,
+					observedOverExpected
+				}
+			}
+		}
+
+		out_data.resolutionsData[resolution] = observedExpectedData
+	}
+
 	return out_data
 
+	// Functions for o,e, o/e
+	async function getObservedValues(file: string, resolution: number, chr1: string, chr2: string) {
+		const observedValues: number[] = []
+
+		// Logic to identify the position of the matrix block for the given chromosomes and resolution
+		const matrixPosition = await getMatrixPosition(file, resolution, chr1, chr2)
+
+		// Read the observed values from the identified matrix block
+		const observedData = isurl
+			? await readHicUrl(file, matrixPosition.position, matrixPosition.length)
+			: await readHicFile(file, matrixPosition.position, matrixPosition.length)
+
+		// Adjust the logic below based on your actual file format
+		const view = new DataView(observedData)
+		let position = 0
+
+		// Example: Reading observed values assuming they are stored as 4-byte floats
```

Comment:
So if this an example, then it's not applicable to all files. This code will need to account for all file versions, data shapes, etc.